pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -0,0 +1,450 @@
+	return phase, err
+}
+
+func phaseFromOutput(plan *planEntry, output string) (string, error) {
+	a := strings.Split(output, "\n")
+	if len(a) < 2 {
+		return "", ErrWaitingf("Could not extract current status from plan for [%s]: status output is incomplete", plan.Machine.Name)
+	}
+	a = strings.Split(a[1], ": ")
+	if len(a) < 2 {
+		return "", ErrWaitingf("Could not extract current status from plan for [%s]: status output is partially complete", plan.Machine.Name)
+	}
+	phase := a[1]
+	logrus.Infof("Rotate Encryption Status [%s]: %s", plan.Machine.Name, phase)
+	return phase, nil
+}
+
+func secretsEncryptInstruction(cp *rkev1.RKEControlPlane, phase string) (plan.OneTimeInstruction, error) {
+	var command string
+
+	switch phase {
+	case Start, ReencryptFinished:
+		command = Prepare
+	case Prepare:
+		command = Rotate
+	case Rotate:
+		command = Reencrypt
+	default:
+		return plan.OneTimeInstruction{}, fmt.Errorf("cannot create secrets-encrypt instruction for node with phase: [%s]", phase)
+	}
+
+	return plan.OneTimeInstruction{
+		Name:    SecretsEncryptApplyCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			command,
+		},
+	}, nil
+}
+
+func statusInstruction(cp *rkev1.RKEControlPlane) plan.OneTimeInstruction {
+	return plan.OneTimeInstruction{
+		Name:    SecretsEncryptStatusCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			"status",
+		},
+		SaveOutput: true,
+	}
+}
+
+func periodicStatusInstruction(cp *rkev1.RKEControlPlane) plan.PeriodicInstruction {
+	return plan.PeriodicInstruction{
+		Name:    SecretsEncryptStatusCommand,
+		Command: rke2.GetRuntimeCommand(cp.Spec.KubernetesVersion),
+		Args: []string{
+			"secrets-encrypt",
+			"status",
+		},
+		PeriodSeconds: 5,
+	}
+}
+
+// restartInstruction generates a restart command for the rke2/k3s server, using the last known leader phase in order to
+// ensure that non-init nodes have a refreshed plan if the leader phase changes. Even if commands were not run on the
+// init node, 'flapping' between restarting the leader and followers is intended until all nodes are on the same status.
+func restartInstruction(cp *rkev1.RKEControlPlane, entry *planEntry, leaderPhase string) plan.OneTimeInstruction {
+	var lastKnownLeaderPhaseEnv string
+	switch entry.Metadata.Labels[rke2.CattleOSLabel] {
+	case windows:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("$env:LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
+	default:
+		lastKnownLeaderPhaseEnv = fmt.Sprintf("LAST_KNOWN_LEADER_PHASE=%s", leaderPhase)
+	}
+	return plan.OneTimeInstruction{
+		Name:    "restart-service",
+		Command: "systemctl",
+		Args: []string{
+			"restart", rke2.GetRuntimeServerUnit(cp.Spec.KubernetesVersion),
+		},
+		Env: []string{
+			lastKnownLeaderPhaseEnv,
+		},
+	}
+}
+
+func (p *Planner) applyAndWait(cp *rkev1.RKEControlPlane, entry *planEntry, node plan.NodePlan) error {
```

Comment:
I'm not sure why this function needs to be defined this way, you can probably just directly call `assignAndCheckPlan`